src/common/providers/identity.ts

```diff
@@ -34,21 +58,263 @@ export type UserRecord = firebase.auth.UserRecord;
+}
+
+/** @internal */
+interface DecodedJwtMetadata {
+  creation_time?: number;
+  last_sign_in_time?: number;
+}
+
+/** @internal */
+interface DecodedJwtUserInfo {
+  uid: string;
+  display_name?: string;
+  email?: string;
+  photo_url?: string;
+  phone_number?: string;
+  provider_id: string;
+}
+
+/** @internal */
+export interface DecodedJwtMfaInfo {
+  uid: string;
+  display_name?: string;
+  phone_number?: string;
+  enrollment_time?: string;
+  factor_id?: string;
+}
+
+/** @internal */
+interface DecodedJwtEnrolledFactors {
+  enrolled_factors?: DecodedJwtMfaInfo[];
+}
+
+/** @internal */
+export interface DecodedJwtUserRecord {
+  uid: string;
+  email?: string;
+  email_verified?: boolean;
+  phone_number?: string;
+  display_name?: string;
+  photo_url?: string;
+  disabled?: boolean;
+  metadata?: DecodedJwtMetadata;
+  password_hash?: string;
+  password_salt?: string;
+  provider_data?: DecodedJwtUserInfo[];
+  multi_factor?: DecodedJwtEnrolledFactors;
+  custom_claims?: any;
+  tokens_valid_after_time?: number;
+  tenant_id?: string;
+  [key: string]: any;
+}
+
+/** @internal */
+export interface DecodedJwt {
+  aud: string;
+  exp: number;
+  iat: number;
+  iss: string;
+  sub: string;
+  event_id: string;
+  event_type: string;
+  ip_address: string;
+  user_agent?: string;
+  locale?: string;
+  sign_in_method?: string;
+  user_record?: DecodedJwtUserRecord;
+  tenant_id?: string;
+  raw_user_info?: string;
+  sign_in_attributes?: {
+    [key: string]: any;
+  };
+  oauth_id_token?: string;
+  oauth_access_token?: string;
+  oauth_refresh_token?: string;
+  oauth_token_secret?: string;
+  oauth_expires_in?: number;
+  [key: string]: any;
+}
+
 /**
  * Helper class to create the user metadata in a UserRecord object
  */
-export class UserRecordMetadata implements firebase.auth.UserMetadata {
+export class UserRecordMetadata implements UserMetadata {
   constructor(public creationTime: string, public lastSignInTime: string) {}
 
   /** Returns a plain JavaScript object with the properties of UserRecordMetadata. */
-  toJSON() {
+  toJSON(): object {
     return {
       creationTime: this.creationTime,
       lastSignInTime: this.lastSignInTime,
     };
   }
 }
 
+/**
+ * Helper class to create the user info in a UserRecord object
+ */
+export class UserRecordInfo implements UserInfo {
+  constructor(
+    public uid: string,
+    public displayName: string,
+    public email: string,
+    public photoURL: string,
+    public providerId: string,
+    public phoneNumber: string
+  ) {}
+
+  toJSON(): object {
+    return {
+      uid: this.uid,
+      displayName: this.displayName,
+      email: this.email,
+      photoURL: this.photoURL,
+      providerId: this.providerId,
+      phoneNumber: this.phoneNumber,
+    };
+  }
+}
+
+/**
+ * Helper class to create the user MultiFactorInfo in a UserRecord object
+ */
+export class UserRecordMultiFactorInfo
+  implements Pick<MultiFactorInfo, keyof MultiFactorInfo> {
```

Comment:
Woah. This is some trippy type decls. Is this how you're saying UserRecordMultiFactorInfo is an element of MultiFactorInfo?